pep-0702.rst

```diff
@@ -0,0 +1,278 @@
+while others (e.g., library developers who want their library to remain
+compatible with future versions of Python) would want to catch any use of
+deprecated functionality in their CI pipeline. Therefore, it is recommended
+that type checker provide configuration options that cover both use cases.
+
+Backwards compatibility
+=======================
+
+Creating a new decorator poses no backwards compatibility concerns.
+As with all new typing functionality, the ``@deprecated`` decorator
+will be added to the ``typing_extensions`` module, enabling its use
+in older versions of Python.
+
+How to teach this
+=================
+
+For users who encounter deprecation warnings in their IDE or type
+checker output, the messages they receive should be clear and self-explanatory.
+Usage of the ``@deprecated`` decorator will be an advanced feature
+mostly relevant to library authors. The decorator should be mentioned
+in relevant documentation (e.g., :pep:`387` and the :exc:`DeprecationWarning`
+documentation) as an additional way to mark deprecated functionality.
+
+Reference Implementation
+========================
+
+TODO
+
+Rejected ideas
+==============
+
+Runtime warnings
+----------------
+
+Users might expect usage of the ``@deprecated`` decorator to issue a
+``DeprecationWarning`` at runtime. However, this would raise a number of
+thorny issues:
+
+* When the decorator is applied to a class or an overload, the warning
+  would not be raised as expected.
+* Users may want to control the ``warn`` call in more detail (e.g.,
+  changing the warning class).
+* ``typing.py`` generally aims to avoid affecting runtime behavior.
+
+Users who want to use ``@deprecated`` while also issuing a runtime warning
+can use the ``if TYPE_CHECKING:`` idiom, for example:
+
+.. code-block:: python
+
+   from typing import TYPE_CHECKING
+   import functools
+   import warnings
+
+   if TYPE_CHECKING:
+       from typing import deprecated
+   else:
+       def deprecated(msg):
+           def decorator(func):
+               @functools.wraps(func)
+               def wrapper(*args, **kwargs):
+                   warnings.warn(msg, DeprecationWarning, stacklevel=2)
+                   return func(*args, **kwargs)
+               wrapper.__deprecated__ = msg
+               return wrapper
+           return decorator
+
+Deprecation of modules and attributes
+-------------------------------------
+
+This PEP covers deprecations of classes, functions and overloads. This
+allows type checkers to detect many but not all possible deprecations.
+To evaluate whether additional functionality would be worthwhile, I
+`examined <https://gist.github.com/JelleZijlstra/ff459edc5ff0918e22b56740bb28eb8b>`__
+all current deprecations in the CPython standard library.
+
+I found:
+
+* 74 deprecations of functions, methods and classes (supported by this PEP)
+* 28 deprecations of whole modules (largely due to :pep:`594`)
```

Comment:
Hm. This does make me think that maybe we should try a bit harder to come up with something for modules. (Maybe at least when they are submodules of a package? Surely that must be the most common case for 3rd party packages.)

Reply:
I feel like the suggested implementation for whole-module deprecations a few paragraphs down is sound, so we can add it if we think it's sufficiently useful. I'll bring this up when I open discussion threads to get more feedback on whether it is useful.

This number might be a bit of a distortion because CPython happens to be in the middle of lots of PEP 594 deprecations, so the number of whole-module deprecations is higher than it would usually be.

Reply:
> This number might be a bit of a distortion because CPython happens to be in the middle of lots of PEP 594 deprecations, so the number of whole-module deprecations is higher than it would usually be.

I actually expect that more ancient, poorly-maintained modules in the stdlib will eventually get the axe.

Reply:
Will bring this up (too) as a point of discussion when starting threads.